File: python/core/tray_optimizer.py
"""
Dragonfly Tray Optimizer
========================
Reduces a standard 120-implant spine tray to the minimum viable set for a
specific case, cutting sterilization cost and OR setup time.

Standard spine tray baseline: 120 implants (industry norm across major vendors).
Dragonfly output: only the sizes required for this patient + safety buffer.

Usage::

    from src.dragonfly.tray_optimizer import TrayOptimizer
    from src.dragonfly.implant_predictor import ImplantPredictor

    plan  = ImplantPredictor().predict(...)
    tray  = TrayOptimizer().optimize(plan, catalog=plan.screws.values())
    print(f"Tray reduced by {tray.reduction_pct:.1f}%")
"""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field

from src.dragonfly.implant_predictor import ImplantPlan, ScrewSpec

# Industry cost estimate per implant unit for sterilization / set maintenance
_STERILIZATION_COST_PER_UNIT_USD = 2.50
_STANDARD_TRAY_BASELINE = 120


# ── Data classes ──────────────────────────────────────────────────────────────

@dataclass
class TrayItem:
    """A single line-item in the optimized tray."""
    diameter_mm: float
    length_mm:   int
    quantity:    int
    role:        str   # "primary" | "backup" | "buffer"

    def label(self) -> str:
        return (f"⌀{self.diameter_mm}mm × {self.length_mm}mm  "
                f"qty={self.quantity}  [{self.role}]")


@dataclass
class TrayConfiguration:
    """
    Optimized implant tray for a single surgical case.

    Attributes
    ----------
    items                       : deduplicated line-items with quantities
    total_implants              : total screw count in optimized tray
    standard_baseline           : always 120 (industry standard tray size)
    reduction_pct               : percentage reduction vs. standard tray
    rod_specs                   : human-readable rod descriptors
    estimated_sterilization_cost_usd : items × $2.50 per unit
    warnings                    : advisory messages from optimizer
    """
    items:                            list[TrayItem]
    total_implants:                   int
    standard_baseline:                int
    reduction_pct:                    float
    rod_specs:                        list[str]
    estimated_sterilization_cost_usd: float
    warnings:                         list[str]


# ── Optimizer ─────────────────────────────────────────────────────────────────

class TrayOptimizer:
    """
    Builds the minimum viable implant tray for a given ImplantPlan.

    Algorithm
    ---------
    For each unique (diameter, length) combination across primary + backup
    screws::

        count = primary_count
              + backup_count
              + (safety_margin × primary_count)   # buffer copies

    Only sizes with count > 0 are included.  The resulting tray is compared
    against the 120-unit industry baseline to compute reduction.
    """

    def optimize(
        self,
        plan:          ImplantPlan,
        catalog:       object = None,   # unused — kept for API compatibility
        safety_margin: int = 2,
    ) -> TrayConfiguration:
        """
        Parameters
        ----------
        plan          : ImplantPlan from ImplantPredictor.predict()
        catalog       : ignored (reserved for future catalog-matching feature)
        safety_margin : extra copies of each primary size added as buffer

        Returns
        -------
        TrayConfiguration
        """
        warnings: list[str] = []

        # ── 1. Count primary screws per (dia, len) ────────────────────────────
        primary_counts: dict[tuple, int] = defaultdict(int)
        for screw in plan.screws.values():
            primary_counts[(screw.diameter_mm, screw.length_mm)] += 1

        # ── 2. Count backup screws per (dia, len) ─────────────────────────────
        backup_counts: dict[tuple, int] = defaultdict(int)
        for backup_list in plan.backup_screws.values():
            for screw in backup_list:
                backup_counts[(screw.diameter_mm, screw.length_mm)] += 1

        # ── 3. Merge into optimized counts ───────────────────────────────────
        # Union of all sizes that appear in primary OR backup sets
        all_sizes: set[tuple] = set(primary_counts) | set(backup_counts)

        items: list[TrayItem] = []
        for (dia, lng) in sorted(all_sizes):
            p_count = primary_counts.get((dia, lng), 0)
            b_count = backup_counts.get((dia, lng), 0)
            buf     = safety_margin * p_count

            total_qty = p_count + b_count + buf
            if total_qty == 0:
                continue

            # Determine dominant role for labelling
            if p_count > 0:
                role = "primary" if b_count == 0 else "primary+backup"
            else:
                role = "backup"

            items.append(TrayItem(
                diameter_mm=dia,
                length_mm=lng,
                quantity=total_qty,
                role=role,
            ))

        total = sum(i.quantity for i in items)

        # ── 4. Reduction vs. standard 120-unit tray ───────────────────────────
        reduction_pct = max(0.0, (1.0 - total / _STANDARD_TRAY_BASELINE) * 100.0)
        if total > _STANDARD_TRAY_BASELINE:
            warnings.append(
                f"Optimized tray ({total} units) exceeds standard baseline "
                f"({_STANDARD_TRAY_BASELINE}); consider splitting into two trays."
            )
            reduction_pct = 0.0

        # ── 5. Rod descriptors ────────────────────────────────────────────────
        rod_specs = _build_rod_specs(plan)

        # ── 6. Sterilization cost estimate ────────────────────────────────────
        sterilization_cost = total * _STERILIZATION_COST_PER_UNIT_USD

        return TrayConfiguration(
            items=items,
            total_implants=total,
            standard_baseline=_STANDARD_TRAY_BASELINE,
            reduction_pct=round(reduction_pct, 1),
            rod_specs=rod_specs,
            estimated_sterilization_cost_usd=round(sterilization_cost, 2),
            warnings=warnings,
        )


# ── Rod spec builder ──────────────────────────────────────────────────────────

def _build_rod_specs(plan: ImplantPlan) -> list[str]:
    """
    Generate human-readable rod descriptor strings.
    Standard construct uses two parallel rods (bilateral).
    """
    mat  = plan.rod_material
    lgth = plan.rod_length_mm
    curv = plan.rod_curvature_deg
    # Bilateral construct — one rod each side
    spec = f"{mat}  {lgth:.0f} mm  {curv:.1f}° pre-bend"
    return [f"Left rod:  {spec}", f"Right rod: {spec}"]

File: unity/Assets/Scripts/Fusion/TrayOptimizer.cs
using System;
using System.Collections.Generic;
using Dragonfly.XR.Domain;

namespace Dragonfly.XR.Fusion
{
    // Reduces a 120-implant standard spine tray to the minimum viable set
    // for a single case. C# port of python/core/tray_optimizer.py.
    public static class TrayOptimizer
    {
        const float SterilizationCostPerUnitUsd = 2.50f;
        const int StandardTrayBaseline = 120;

        public static TrayConfiguration Optimize(ImplantPlan plan, int safetyMargin = 2)
        {
            var warnings = new List<string>();

            var primaryCounts = new Dictionary<(float dia, int len), int>();
            foreach (var screw in plan.Screws.Values)
            {
                var k = (screw.DiameterMm, screw.LengthMm);
                primaryCounts.TryGetValue(k, out int n);
                primaryCounts[k] = n + 1;
            }

            var backupCounts = new Dictionary<(float dia, int len), int>();
            foreach (var list in plan.BackupScrews.Values)
            {
                foreach (var screw in list)
                {
                    var k = (screw.DiameterMm, screw.LengthMm);
                    backupCounts.TryGetValue(k, out int n);
                    backupCounts[k] = n + 1;
                }
            }

            var allSizes = new HashSet<(float dia, int len)>(primaryCounts.Keys);
            foreach (var k in backupCounts.Keys) allSizes.Add(k);

            var sizes = new List<(float dia, int len)>(allSizes);
            sizes.Sort((a, b) =>
            {
                int c = a.dia.CompareTo(b.dia);
                return c != 0 ? c : a.len.CompareTo(b.len);
            });

            var items = new List<TrayItem>();
            foreach (var k in sizes)
            {
                primaryCounts.TryGetValue(k, out int p);
                backupCounts.TryGetValue(k, out int b);
                int buf = safetyMargin * p;
                int qty = p + b + buf;
                if (qty == 0) continue;

                string role = p > 0
                    ? (b == 0 ? "primary" : "primary+backup")
                    : "backup";

                items.Add(new TrayItem
                {
                    DiameterMm = k.dia,
                    LengthMm = k.len,
                    Quantity = qty,
                    Role = role,
                });
            }

            int total = 0;
            foreach (var i in items) total += i.Quantity;

            float reductionPct = Math.Max(0f, (1f - (float)total / StandardTrayBaseline) * 100f);
            if (total > StandardTrayBaseline)
            {
                warnings.Add(
                    $"Optimized tray ({total} units) exceeds standard baseline " +
                    $"({StandardTrayBaseline}); consider splitting into two trays.");
                reductionPct = 0f;
            }

            return new TrayConfiguration
            {
                Items = items,
                TotalImplants = total,
                StandardBaseline = StandardTrayBaseline,
                ReductionPct = (float)Math.Round(reductionPct, 1),
                RodSpecs = BuildRodSpecs(plan),
                EstimatedSterilizationCostUsd =
                    (float)Math.Round(total * SterilizationCostPerUnitUsd, 2),
                Warnings = warnings,
            };
        }

        static List<string> BuildRodSpecs(ImplantPlan plan)
        {
            string spec =
                $"{plan.RodMaterial}  {plan.RodLengthMm:F0} mm  " +
                $"{plan.RodCurvatureDeg:F1}° pre-bend";
            return new List<string>
            {
                $"Left rod:  {spec}",
                $"Right rod: {spec}",
            };
        }
    }
}
